content/blogs/gui-setup-part3-first-testing-problems.md

```diff
@@ -0,0 +1,267 @@
+
+These two steps are the bottleneck of this whole file-copy loop:
+because, for each file to copy, potentially present in the _same_ CAB archive, the archive is opened, the file has to be found (by looping through them to find the correct one) and extracted, then the CAB archive is closed.
+And this operation is done over and over again.
+
+In comparison, it is _much faster_ to open the CAB archive _only once_, then whenever a file needs to be copied, check whether it exists in the CAB archive, starting from the current search position, and if possible, exploit the fact that the files in the copy queue are sorted in order of their appearance in the CAB archive.
+(Alternatively, loop once through all the files in the CAB archive and check whether each successive of these are files that need to be extracted.)
+The exception could be, that if the next file is not in the order, one restarts finding from the beginning of the CAB archive.
+This would be the slow code path, but not actually taken in real life due to the reasons from above.
+
+This is exactly what the text-mode USETUP installer does, and why ReactOS installation time is there blazing fast in comparison.
+
+Additionally, it appears, after some further investigation (suggested by Mark Jansen), that each read operation from the CAB file, implemented with a `fdi->read()` callback (registered when initializing the fdi context with the `FDICreate()` function, done by setupapi), that does a `ReadFile()` call.
+Meaning, basically, that each data read from the CAB triggers a file I/O operation to disk, making the whole operation slower than it could be.
+This could be alleviated by adding ReactOS-specific code in setupapi, to memory-map the CAB archive instead so that reading from it can be done directly via memory reads.
+Such a thing is also done by the text-mode USETUP.
+
+#### The current workaround
+
+To alleviate the problems described above, the simplest workaround is to just use USETUP's file-copy and extraction functions. With minimal adaptations the existing code can be easily adapted for usage in the GUI installer. File copy works and finishes speedily.
+
+{{< gallery >}}
+{{< figure link="/img/blogs/hbelusca_gui-setup/guisetup_filecopy_finish.png" src="/img/blogs/hbelusca_gui-setup/guisetup_filecopy_finish.png" caption="Last file being copied. The directory listing shows that the files are there.">}}
+{{< /gallery >}}
+
+
+### 3rd bug
+
+At the end of the installation, the ReactOS GUI installer creates the necessary registry hives (SYSTEM, SOFTWARE, SAM and SECURITY) in their correct place in the installation directory (`%SystemRoot%\system32\config`).
+Using NT APIs, they are created, filled for one part with default settings and for another part with installation-specific settings, then flushed back to disk.
+
+Notice that these created registry hives are separate from the system root ones being used by the LiveCD.
```

Comment:
```suggestion
Notice that the created registry hives are separate from those being used by the LiveCD system itself.
```

Reply:
"LiveCD system itself" , or "LiveCD itself" ?

Reply:
I think "LiveCD itself" would be fine too.

Reply:
ok this is what I have implemented.